cpmpy/solvers/rc2.py

```diff
@@ -0,0 +1,298 @@
+
+    ===============
+    List of classes
+    ===============
+
+    .. autosummary::
+        :nosignatures:
+
+        CPM_rc2
+
+    ==============
+    Module details
+    ==============
+"""
+from .solver_interface import SolverStatus, ExitStatus
+from .pysat import CPM_pysat
+from ..exceptions import NotSupportedError
+from ..expressions.variables import _BoolVarImpl, _IntVarImpl, NegBoolView
+from ..transformations.linearize import only_positive_coefficients_
+from ..transformations.get_variables import get_variables
+from ..transformations.flatten_model import flatten_objective
+from ..transformations.int2bool import get_user_vars, _encode_lin_expr
+
+
+class CPM_rc2(CPM_pysat):
+    """
+    Interface to PySAT's RC2 MaxSAT solver API.
+
+    Creates the following attributes (see parent constructor for more):
+
+    - ``pysat_vpool``: a pysat.formula.IDPool for the variable mapping
+    - ``pysat_solver``: a pysat.examples.rc2.RC2() (or .RC2Stratified())
+    - ``ivarmap``: a mapping from integer variables to their encoding for `int2bool`
+    - ``encoding``: the encoding used for `int2bool`, choose from ("auto", "direct", "order", or "binary"). Set to "auto" but can be changed in the solver object.
+
+    The :class:`~cpmpy.expressions.globalconstraints.DirectConstraint`, when used, calls a function on the ``pysat_solver`` object.
+
+    Documentation of the solver's own Python API:
+    https://pysathq.github.io/docs/html/api/examples/rc2.html
+
+    .. note::
+        CPMpy uses 'model' to refer to a constraint specification,
+        the PySAT docs use 'model' to refer to a solution.
+
+    """
+
+    def __init__(self, cpm_model=None, subsolver=None):
+        """
+        Constructor of the native solver object
+
+        Requires a CPMpy model as input, and will create the corresponding
+        PySAT clauses and solver object
+
+        Only supports optimisation problems (MaxSAT)
+
+        Arguments:
+            cpm_model (Model(), optional): a CPMpy Model()
+            subsolver (None): ignored
+        
+        """
+        if not self.supported():
+            raise ImportError("PySAT is not installed. The recommended way to install PySAT is with `pip install cpmpy[pysat]`, or `pip install python-sat` if you do not require `pblib` to encode (weighted) sums.")
+
+        from pysat.formula import IDPool, WCNFPlus, CNFPlus
+
+        # In order for all features to work (e.g. `process` option), RC2 should be bootstrapped from a formula
+        self.pysat_solver = WCNFPlus()
+
+        def append_formula(cnf, weight=None):
+            """Some RC2 oracles/solvers support AtMostK constraints in the form of CNFPlus objects, which can be passed through this function in addition to regular clauses."""
+            clauses = []
+            atmosts = []
+            if isinstance(cnf, list):
+                clauses = cnf
+            elif isinstance(cnf, tuple):
+                atmosts = cnf
+            else:  # CNFPlus
```

Comment:
is this expected? can people given an entire CNFPlus to us? or to pysat for that manner?
how would you typehint this `add_formula`... and is the typehint the same as the original `add_formula` function? If not, this is imho not right...